### PR TITLE
V1 DataGenerator and Data Augment Revamp on GPU ( no unittest updates)

### PR DESCRIPTION
**Pull request by Thelukepet**
_(continuation from cv2-tfm-replacement branch)_
- Refactored CPU (cv2) / deprecated image manipulation packages (TFA)

- Refactored data augmentations to be part of the final Sequential model for GPU support, this happens within the `customize_model` function of `management.py`
   - Dynamic handling of subsequent data augments and different number of channels (1, 3 or 4)
   - Custom keras Layer classes for each type of augmentation, can be easily extended in the future with new augments
   - Data augmentation visualizations with --visualize_augments on the test images within `tests/data` (will be placed in the visualize folder in the future)
   - Random augment model creation with --aug_random_augments for debugging purposes (will be removed later)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -90,6 +90,24 @@ The command-line options include, but are not limited to:
...
+  - `--aug_random_width`: Adjusts image width randomly and maintains original dimensions by either
+        compressing or padding the image. 
+  - `--aug_binarize_otsu`: Binarize the input tensor using Otsu's tresholding
+  - `--aug_binarize_sauvola`: Binarize the input tensor using Sauvola's 
+    thresholding (**warning**: computationally expensive to run)
+  - `--aug_random_shear`: Randomly shear each image along the X-axis, i.e. 
+    randomly slant the image to the right or left
+  - `--aug_blur_image`: Randomly apply a Gaussian blur over each image (will 
+    be made milder if binarization is active to prevent smudged results)
+  - `--aug_invert`: Invert the pixels in the image
```
Comment (TimKoornstra) [inline README.md]: Ik zou die `--aug_` lekker weglaten, aangezien dat al eerder supported was. Dat scheelt weer mensen die zich afvragen wat er mis gaat met deze update & het is wel duidelijk dat dit augmentaties zijn

File: src/data/data_handling.py [hunk 2]
```diff
@@ -54,19 +54,19 @@ def initialize_data_loader(args: argparse.Namespace, char_list: List[str],
         test_list=args.test_list,
         inference_list=args.inference_list,
         char_list=char_list,
-        do_binarize_sauvola=args.do_binarize_sauvola,
-        do_binarize_otsu=args.do_binarize_otsu,
+        aug_binarize_sauvola=args.aug_binarize_sauvola,
+        aug_binarize_otsu=args.aug_binarize_otsu,
```
Comment (TimKoornstra) [inline src/data/data_handling.py]: zelfde als bij requirements.txt

File: src/data/generator.py [hunk 3]
```diff
@@ -21,202 +16,40 @@ def __init__(self,
...
-                 random_width=False,
-                 distort_jpeg=False,
                  channels=1,
-                 do_random_shear=False,
-                 do_blur=False,
-                 do_invert=False
+                 aug_binarize_sauvola=False,
+                 aug_binarize_otsu=False,
+                 aug_elastic_transform=False,
+                 aug_random_crop=False,
```
Comment (TimKoornstra) [inline src/data/generator.py]: .

File: src/data/loader.py [hunk 4]
```diff
@@ -111,20 +111,20 @@ def generators(self):
...
-                        'do_binarize_otsu': self.do_binarize_otsu,
-                        'do_elastic_transform': self.elastic_transform,
-                        'random_crop': self.random_crop,
-                        'random_width': self.random_width,
-                        'distort_jpeg': self.distort_jpeg,
-                        'do_random_shear': self.do_random_shear
+                        'aug_binarize_sauvola': self.aug_binarize_sauvola,
+                        'aug_binarize_otsu': self.aug_binarize_otsu,
+                        'aug_elastic_transform': self.aug_elastic_transform,
+                        'aug_random_crop': self.aug_random_crop,
```
Comment (TimKoornstra) [inline src/data/loader.py]: .

File: src/model/management.py [hunk 5]
```diff
@@ -5,11 +5,14 @@
...
 from typing import Any, List, Dict
 
+import keras
+import matplotlib.pyplot as plt
 # > Third-party dependencies
 import tensorflow as tf
 
 # > Local dependencies
 from utils.utils import load_model_from_directory
+from data.augment_layers import *
```
Comment (TimKoornstra) [inline src/model/management.py]: Liefst geen * imports gebruiken. Dit heeft meerdere redenen:

1. **Alles** wordt geimporteerd. Dit kan tot conflicten leiden als je in 2 files/imports dezelfde function/variable names hebt.
2. Lastig te lezen waar een functie vandaan komt voor mensen die je code base niet kennen.
3. Je hebt niet altijd een hele module nodig, wat kan leiden tot slechtere performance.

File: src/model/management.py [hunk 6]
```diff
@@ -119,9 +121,127 @@ def customize_model(model: tf.keras.Model, args: argparse.Namespace,
...
+            args.aug_binarize_sauvola, args.aug_blur, args.aug_invert]):
+
+        # Set input params from trainings model input spec
+        batch_size, width, height, channels = (model.layers[0]
+                                               .get_input_at(0)
+                                               .get_shape().as_list())
+
+        # Retrieve the (random) augmentation model from the selected augments
+        augment_options = get_augment_classes()
+        augment_selection = get_augment_model()
```
Comment (TimKoornstra) [inline src/model/management.py]: Je kunt het beste de args hier meegeven in plaats van in `augment_layers` opnieuw gaan getten. Dan doe je dat _en_ dubbel _en_ snap ik niet hoe de options en selections bepaald worden zonder naar die file te gaan.

File: src/model/management.py [hunk 7]
```diff
@@ -119,9 +121,127 @@ def customize_model(model: tf.keras.Model, args: argparse.Namespace,
...
+                save_augment_steps_plot(
+                    aug_model,
+                    sample_image_path="../tests/data/test-image"
+                                      + str(img_num)
+                                      + ".png",
+                    save_path="augment_test_img_"
+                              + str(img_num)
+                              + ".png",
+                    channels=args.channels)
+            logging.info("Augment visualizations are stored in the src folder")
```
Comment (TimKoornstra) [inline src/model/management.py]: Misschien dat je dit zelfs gewoon als een `if __name__ == "__main__"` onderaan deze file kunt zetten. Anders moet je steeds het hele script opstarten, etc. voordat je je kleine aanpassing kunt bekijken.

Comment (TimKoornstra) [reply src/model/management.py]: En t scheelt weer een arg ;)

File: src/model/management.py [hunk 8]
```diff
@@ -119,9 +121,127 @@ def customize_model(model: tf.keras.Model, args: argparse.Namespace,
...
+    after binarization
+    """
+    if background_color is None:
+        background_color = [1, 1, 1]
+    rgb = image[..., :3]
+    alpha = tf.expand_dims(image[..., 3], axis=-1)
+    return rgb * alpha + background_color * (alpha - 1)
+
+
+def save_augment_steps_plot(aug_model, sample_image_path, save_path, channels):
```
Comment (TimKoornstra) [inline src/model/management.py]: Waar zijn de type hints :( hieronder staan ze wel

File: src/setup/arg_parser.py [hunk 9]
```diff
@@ -119,16 +119,28 @@ def get_arg_parser():
     augmentation_args = parser.add_argument_group('Augmentation arguments')
     augmentation_args.add_argument('--augment', action='store_true',
                                    help='beta: apply data augmentation to training set. In general this is a good idea')
-    augmentation_args.add_argument('--elastic_transform', action='store_true',
+    augmentation_args.add_argument('--aug_elastic_transform',
+                                   action='store_true',
                                    help='beta: elastic_transform, currently disabled')
-    augmentation_args.add_argument('--random_crop', action='store_true',
+    augmentation_args.add_argument('--aug_random_crop', action='store_true',
```
Comment (TimKoornstra) [inline src/setup/arg_parser.py]: .

File: tests/test_data_augments.py
Comment (TimKoornstra) [inline tests/test_data_augments.py:1]: Ik zou hier graag nog een aantal tests willen zien die verschillende combinaties van augments testen

File: tests/test_datagenerator.py
Comment (TimKoornstra) [inline tests/test_datagenerator.py:1]: Weinig tests, maar ik keur m goed omdat deze class toch binnenkort weg gaat

File: src/data/augment_layers.py [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+            shape=[1, 8]
+        )
+
+        # Apply the shear transformation
+        # Fill value is set to 0 to ensure that binarization is not affected
+        sheared_image = tf.raw_ops.ImageProjectiveTransformV3(
+            images=inputs,
+            transforms=shear_matrix_tf,
+            output_shape=output_shape,
+            fill_value=1,
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: Maar fill value is 1?

File: src/data/augment_layers.py [hunk 11]
```diff
@@ -0,0 +1,600 @@
...
+        ----------
+        - inputs (tf.Tensor): The original image tensor to be transformed.
+        Can handle tensors of various shapes and channel numbers (e.g.,
+        grayscale, RGB, RGBA).
+
+        Returns
+        -------
+        - tf.Tensor: The image tensor after applying elastic transformation.
+
+        Notes ----- - The `etf.deform_grid` function is used to apply the
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: oopsie

File: src/data/augment_layers.py [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+
+        Returns:
+        - tf.Tensor: Resized and padded image tensor.
+        """
+        # Calculate the padding sizes
+        padding_width = self.additional_width
+        pad_left = padding_width // 2
+        pad_right = padding_width - pad_left
+
+        # Apply "white" padding
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: Waarom niet zwart?

File: src/data/augment_layers.py [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+                return ((image > thresh) * 1).astype(np.float32)
+
+            logging.info("WARNING: sauvola is extremely inefficient, select "
+                         "Otsu for faster processing")
+            binarized = tf.numpy_function(sauvola_thresh, [processed_input],
+                                          tf.float32)
+
+        else:
+            raise ValueError(
+                "Invalid method. Supported methods are 'otsu' and 'sauvola'.")
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: Dit kun je beter in de init zetten

File: src/data/augment_layers.py [hunk 14]
```diff
@@ -0,0 +1,600 @@
...
+            binarized = tf.numpy_function(otsu_thresh, [processed_input],
+                                          tf.float32)
+
+        elif self.method == 'sauvola':
+            def sauvola_thresh(image):
+                thresh = threshold_sauvola(image, window_size=self.window_size)
+                return ((image > thresh) * 1).astype(np.float32)
+
+            logging.info("WARNING: sauvola is extremely inefficient, select "
+                         "Otsu for faster processing")
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: logging.warning

File: src/data/augment_layers.py [hunk 15]
```diff
@@ -0,0 +1,600 @@
...
+
+        Parameters:
+        - inputs (tf.Tensor): Input image tensor.
+
+        Returns:
+        - tf.Tensor: Blurred image tensor.
+
+        Parameters
+        ----------
+        inputs : object
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: 2x parameters O.o

File: src/data/augment_layers.py [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+            left_pad = padding_width // 2
+            right_pad = padding_width - left_pad
+
+            # Pad the new image to retain overall width for compatibility
+            padded_image = tf.pad(resized_image,
+                                  paddings=[[0, 0], [0, 0],
+                                            [left_pad, right_pad],
+                                            [0, 0]],
+                                  mode="CONSTANT",
+                                  constant_values=1)
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: waarom value 1?

File: src/data/augment_layers.py [hunk 17]
```diff
@@ -0,0 +1,600 @@
...
+        # exclude invert for random augment selection
+        if name.startswith("Invert"):
+            continue
+        # exclude padding layer for random augment selection
+        if name.startswith("ResizeWithPadLayer"):
+            continue
+        elif isinstance(obj, type):
+            layer_instance = obj()
+            augment_options.append(layer_instance)
+    return augment_options
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: Ik weet niet wat ik vind van deze functie

Comment (TimKoornstra) [reply src/data/augment_layers.py]: Globals kopiëren, etc. Waarom niet gewoon de lijst meegeven? En de naam is een beetje misleidend, het zouden eerder instances moeten zijn.

File: src/data/augment_layers.py [hunk 18]
```diff
@@ -0,0 +1,600 @@
...
+    require some additional processing layers to ensure that information is
+    not lost from the origian image.
+
+    Returns:
+    list
+        A list of custom data augment layers that inherit from
+        tf.keras.layers.Layer
+    """
+    logger = logging.getLogger(__name__)
+    args = get_args()
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: Gewoon args passen aan de functie. Dit is niet efficient en onhandig

File: src/data/augment_layers.py [hunk 19]
```diff
@@ -0,0 +1,600 @@
...
+        augmentation layers.
+
+    """
+
+    # Check if both types of binarization are present, remove Sauvola if true
+    otsu_present = any(isinstance(obj, BinarizeLayer)
+                       and obj.method == 'otsu' for obj in augment_selection)
+    sauvola_present = any(isinstance(obj, BinarizeLayer)
+                          and obj.method == 'sauvola' for obj in
+                          augment_selection)
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: Ik denk dat het efficienter kan dan 2x door de lijst loopen. En de naam obj is een beetje vaag; misschien gewoon `layer`? of `aug_layer`?

File: src/data/augment_layers.py [hunk 20]
```diff
@@ -0,0 +1,600 @@
...
+        # Remove the object of type BinarizeLayer with Sauvola method
+        augment_selection = [obj for obj in augment_selection
+                             if not (isinstance(obj, BinarizeLayer)
+                                     and obj.method == 'sauvola')]
+
+    # Init blur params
+    mild_blur = False
+    blur_index = -1
+
+    # Check if a blur occurs before binarization, if so replace with mild blur
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: Waarom?

File: src/data/augment_layers.py [hunk 21]
```diff
@@ -0,0 +1,600 @@
...
+    for i, augment in enumerate(augment_selection):
+        if isinstance(augment, BinarizeLayer):
+            binarize = True
+            binarize_index = i
+        elif binarize and (isinstance(augment, BlurImageLayer)):
+            # Remove BlurImageLayer that occurs after binarization
+            augment_selection.remove(augment)
+            # Insert a mild blur object right before binarize
+            augment_selection.insert(binarize_index,
+                                     BlurImageLayer(mild_blur=True))
```
Comment (TimKoornstra) [inline src/data/augment_layers.py]: Werkt dit? Lijkt me niet dat je iets mag inserten in een lijst waar je op dat moment doorheen aan het loopen bent. Als het wel kan, lijkt me alleen maar dat er nare side-effects van zullen komen.

Comment (TimKoornstra) [reply src/data/augment_layers.py]: Misschien beter om de loop te doen met `for i, augment in enumerate(augment_selection[:]):`

**Review by TimKoornstra — CHANGES_REQUESTED** [review]
Ziet er goed uit :) Paar dingetjes die beter kunnen, maar verder ziet t er heel goed uit. Maarja ik heb t ook nog niet getest haha

**Review by TimKoornstra — APPROVED** [review]

File: src/model/management.py (hunk 5 above)
Comment (Thelukepet) [reply src/model/management.py]: Eens, alleen heb ik binnen data.augment_layers altijd wel alle classes nodig die erin zitten. Normaliter probeer ik dit ook zo min mogelijk te gebruiken maar in dit geval kreeg ik anders een erg lelijke import met 9 classes

File: src/data/augment_layers.py (hunk 10 above)
Comment (Thelukepet) [reply src/data/augment_layers.py]: Fout in comment, het moest 1 zijn

File: src/data/augment_layers.py (hunk 12 above)
Comment (Thelukepet) [reply src/data/augment_layers.py]: Zwarte padding zorgt ervoor dat binarization kapot gaat omdat de functie dan de waardes van de padding meeneemt

File: src/data/augment_layers.py (hunk 16 above)
Comment (Thelukepet) [reply src/data/augment_layers.py]: Ook weer vanwege binarization, zwarte padding zorgt voor problemen, uiteraard geen probleem als de operatie post-binarizatie wordt toegepast

File: src/data/augment_layers.py (hunk 17 above)
Comment (Thelukepet) [reply src/data/augment_layers.py]: Ik pak specifiek de globals uit de file omdat ik dan zeker weet dat als er nieuwe layers zijn toegevoegd dat deze ook worden meegenomen. Dit is gedaan zodat ik niet telkens een lijst hoef up te daten, maar ik snap dat je een oplossing met een lijst logischer / beter te begrijpen vindt

File: src/data/augment_layers.py (hunk 20 above)
Comment (Thelukepet) [reply src/data/augment_layers.py]: Eerste twee lines zijn puur voor initialisatie, ik vervang de blur met een "milde blur" omdat anders het resultaat compleet onleesbaar wordt na binarisatie.
